bootstrap/utils.go

```diff
@@ -134,11 +135,23 @@ func downloadAndUnzip(url string, path string, target string) error {
 	return unzip(path, target)
 }
 
-func executeCmd(cmd string, args []string) (*exec.Cmd, error) {
+// Executes a command, waits for it to finish and returns any errors.
+func executeCmd(cmd string, args []string) error {
 	command := exec.Command(cmd, args...) // #nosec
-	err := command.Start()
-	if err != nil {
-		return nil, err
-	}
-	return command, nil
+	return command.Run()
+}
+
+// Executes a command in the background. Returns a context so that the caller may cancel the
+// command prematurely if necessary, as well as an errors channel.
+func executeBackgroundCmd(cmd string, args []string) (context.CancelFunc, <-chan error) {
```

Comment:
Interesting. Nice job for catching this. I wouldn't jump to blaming the Golang std library for having a bug so fast, though. To me it makes more sense that we're simply doing something wrong...

I wanted to use `context` since it was built exactly for these scenarios (chaining calls and cancelling them properly when something happens). It's definitely possible to kill the process, however this feels to me like the equivalent of flipping on a light switch, then breaking the light bulb to turn off the light instead of flipping the switch back off again :-) Hope that makes sense.

Let me review this behavior to try and pinpoint the issue with using the context library (unless you think it's wrong to use it here, assuming we could get it to kill the process properly).

Reply:
Also worth noting that in the end the call to `cancel()` just sends `SIGKILL` so it's the same thing.
> The provided context is used to kill the process (by calling os.Process.Kill) if the context becomes done before the command completes on its own.

Reply:
True, the end result is the same when cancelling or killing, I just felt that it's more correct Golang-wise to start a background task, then cancel it "properly" instead of killing it "from the outside" if that makes sense.

Reply:
I'm still investigating this, but just wanted to say in the meantime - `cmd.Start()` must be followed by `cmd.Wait()`, otherwise the command's resources aren't getting freed according to the [docs](https://golang.org/pkg/os/exec/#Cmd.Start).

Inspection of the source code for `Run()` shows that it is actually `Start()` followed by `Wait()`:

```go
func (c *Cmd) Run() error {
	if err := c.Start(); err != nil {
		return err
	}
	return c.Wait()
}
```

Therefore, I am not sure the test you're performed using `Start()` is reliable. I am, however, indeed seeing that when running a child process which ignores SIGINT, the child is left behind when sending SIGINT to the main process. I'm getting the feeling this has something to do with the main process exiting and not letting `defer`s complete. Updates will follow.

Reply:
OK, I think I figured out the problem. At the end of [Start()](https://golang.org/src/os/exec/exec.go?s=10035:10062#L328) there is the following:

```go
if c.ctx != nil {
    c.waitDone = make(chan struct{})
    go func() {
        select {
        case <-c.ctx.Done():
            c.Process.Kill()
        case <-c.waitDone:
        }
    }()
}
```

The `Start()` function blocks until either the channel returned by `c.ctx.Done()` is closed (explicit cancelation of the context) **or** until `c.waitDone` is closed (command exited on its own).

When the main process returns, it runs the deferred `cancel()` which closes `c.ctx.Done()`, however the main process then **immediately terminates** before `c.Process.Kill()` had a chance to run. This happens because `cancel()` doesn't block until the kill function returns, thus creating a race condition between the main function and the kill function. If the main function wins, the kill function never runs and the child process is leaked.

The solution:

```go
func executeBackgroundCmd(wg *sync.WaitGroup, cmd string, args ...string) (context.CancelFunc, <-chan error) {
	ctx, cancel := context.WithCancel(context.Background())
	command := exec.CommandContext(ctx, cmd, args...)

	errChan := make(chan error, 1)
	go func() {
		defer wg.Done()
		err := command.Run()
		errChan <- err
	}()

	return cancel, errChan
}

func main() {
	var wg sync.WaitGroup

	wg.Add(1)
	cancel, errChan := executeBackgroundCmd(&wg, "./proctest/proctest")
	defer func() {
		cancel()
		wg.Wait()
	}()

	stop := make(chan os.Signal, 1)
	signal.Notify(stop, syscall.SIGINT, syscall.SIGTERM)

	select {
	case <-stop:
		log.Println("Got stop signal")
	case err := <-errChan:
		log.Printf("Got error: %v", err)
	}
}
```

This seems to work fine.